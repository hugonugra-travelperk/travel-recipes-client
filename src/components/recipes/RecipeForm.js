import React, { useState } from "react";
import {useNavigate} from "react-router-dom";
import {CardWrapper, CardHeader, CardBody} from '../../components/base/Card'
import {Button} from '../../components/base/Button'
import {InputForm} from '../../components/base/Input'
import IngredientsList from "../../components/ingredients/IngredientsList";
import useInput from '../../hooks/useInput'
import useAxios from '../../hooks/useAxios'

export default function RecipeForm(props) {
  let navigate = useNavigate();

  const DOMAIN_URL = '/recipes/'

  const [name, nameInput, resetName] = useInput({
    type: 'text',
    placeholder: 'Name',
    initial_value: props.recipe.name
  });

  const [description, descriptionInput, resetDescription] = useInput({
    type: 'area',
    placeholder: 'Description',
    initial_value: props.recipe.description
  });

  const [new_ingredient_name, newIngredientNameInput, resetIngredientName] = useInput({
    type: 'text',
    placeholder: 'New Ingredient',
  });
  
  const [ingredients, setIngredients] = useState(props.recipe.ingredients);

  const url = ('id' in props.recipe) ? DOMAIN_URL + props.recipe.id + '/' : DOMAIN_URL  

  const { dispatch, response, loading, error } = useAxios({
    method: props.method,
    url,
    data: {
      name,
      description,
      ingredients,
    }
  }, performAction)

  function add_ingredient() {
    setIngredients([...ingredients, {'name': new_ingredient_name }])
    resetIngredientName()
  }

  function performAction(res, err) {
    if(err === null && res) {
      setTimeout(()=>{navigate('/')}, 1000) 
    }
  }

  function goBack() {
    navigate('/')
  }

  return (
    <CardWrapper>
      <CardHeader>
        {props.title}
      </CardHeader>
      {error}
      <CardBody>
        <InputForm>
          {nameInput}
          {descriptionInput}
          Ingredients
          <IngredientsList ingredients={ingredients}/>
          {newIngredientNameInput}
          <Button primary onClick={() => add_ingredient()}> Add </Button>  
        </InputForm>
        <Button primary onClick={() => dispatch()}> {props.title} </Button>
        <Button primary onClick={() => goBack()}> Back </Button>
      </CardBody>
    </CardWrapper>
  )
}
