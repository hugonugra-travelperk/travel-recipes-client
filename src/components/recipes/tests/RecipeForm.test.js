import {render, cleanup} from "@testing-library/react"
import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import RecipeForm from "../RecipeForm"

afterEach(cleanup);

const recipe = {'id':'1', 'name': 'test text recipe',
          'ingredients': [{'id': '1', 'name': 'test text ingredient'}]
        }

it("renders", () => {
  const {asFragment} = render(
    <BrowserRouter>
      <RecipeForm title="New Recipe" method='POST' recipe={recipe}/>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
