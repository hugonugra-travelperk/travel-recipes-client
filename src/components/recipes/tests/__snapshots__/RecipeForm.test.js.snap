// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL gaBRgU"
  >
    <header
      class="sc-gsDKAQ jGoJmq"
    >
      New Recipe
    </header>
    <div
      class="sc-dkPtRN PHDyg"
    >
      <div
        class="sc-gKclnd fVLuFs"
      >
        <input
          class="sc-iCfMLu gtbZOl"
          placeholder="Name"
          type="text"
          value="test text recipe"
        />
        <textarea
          class="sc-furwcr eoRGTG"
          cols="30"
          placeholder="Description"
          rows="3"
          type="area"
        />
        Ingredients
        <div
          class="sc-eCImPb gayBGf"
        >
          <ul>
            <li
              class="sc-pVTFL eOLXxe"
            >
              <span
                class="sc-jrQzAO cOnMjS"
              >
                test text ingredient
              </span>
            </li>
          </ul>
        </div>
        <input
          class="sc-iCfMLu gtbZOl"
          placeholder="New Ingredient"
          type="text"
          value=""
        />
        <button
          class="sc-jRQBWg cMuxuo"
        >
           Add 
        </button>
      </div>
      <button
        class="sc-jRQBWg cMuxuo"
      >
         New Recipe 
      </button>
      <button
        class="sc-jRQBWg cMuxuo"
      >
         Back 
      </button>
    </div>
  </div>
</DocumentFragment>
`;
